### ISSUE TITLE
FlowCollectionLogEntry still giving null pointer exception

### ISSUE THREAD

**Issue by vr8hub** [jongpie/NebulaLogger#213]
### Package Edition of Nebula Logger

Unlocked Package

### Package Version of Nebula Logger

v4.6.8

### New Bug Summary

I had 4.6.6 installed. I upgraded to 4.6.8. I've confirmed in the Installed Packages window that it says 4.6.8 for Logger. However, just in case it matters, the flow with the log actions was created while 4.6.6 was installed. (I'm assuming/hoping that doesn't matter, just mentioning it in case.)

I know the release notes say that the change was made, and I see the change to change the parameter to not required, but I'm still getting a null pointer exception from the FlowCollectionLogEntry method if the record collection parameter is NULL.

From the flow debug log:
````
FLOWCOLLECTIONLOGENTRY (APEX): Log End
Inputs:
saveLog = true
flowName = CAN_MedicalVisitAfterUpdUpdateRptStat
message = Updating RptColl
records = {!RptStatColl} ([null])

Error Occurred: An Apex error occurred: System.NullPointerException: Attempt to de-reference a null object 
````

Is it possible the code that _processes_ the records wasn't changed to look for an empty list before trying to handle it? Note: 1) I am not a developer nor do I play one on TV, but it seems like the likeliest scenario, and 2) I didn't see at first glance any changes to the processing code, but there's a lot of code and see 1. :)

**Comment by jongpie** [comment]
Hi @vr8hub - sorry that there are still some issues with this! I'm trying to reproduce the issue, but having some problems doing so - when I log a null record collection in Flow, everything seems to work successfully for me. However, based on the log you included, it looks like you're possibly logging a record collection that contains a null record (as opposed to the record collection itself being null). Can you share some details on how you've built your Flow? Any details around how you're creating the record collection would be helpful. I _think_ I know how to resolve the code's issue, but I'd like to be able to reproduce the issue first before I make any changes.

Thanks again for your help!

**Comment by vr8hub** [comment]
Well, I didn't think I was (this was actually one of the reasons I was logging, to try to figure out what was going on), but as it turns out, it appears I was. So this one's on me. So sorry! And perfectly understandable if you can't support every possible error we make. :)

There are about four different paths in the flow to get to the point where a record variable is added to the collection, and in three of them, the record has something in it. But I missed on the fourth one, and in that case the record is empty when it's added to the collection.

Mea culpa!

**Comment by vr8hub** [comment]
Well, if you _can_ support it, it would be great; logging is a good way to find bad or missing logic, so for record and record collection logging, whether they're empty or not, being able to log everything would be a huge help. Thanks, Jonathan!

**Comment by jongpie** [comment]
Sounds good, I definitely think I can handle this scenario - I'll keep you updated on my progress!

**Comment by jongpie** [comment]
@vr8hub good news - I was able to fix this issue fairly easily, as well as some other related bugs! The next release, v4.7.9, will include changes for these scenarios:
1. **Fixed** - Logging failed when logging a record collection (`List<SObject>`) where the first record was null
2. **Fixed** - Logging failed when logging an empty record collection
3. **Improved** - Logging "worked" when logging a null record or null record collection, but there was no clear indication on the `LogEntry__c` record that you had even tried to log a record/record collection. Now, it will log the word "null" in the field `RecordJson__c`, and "Unknown" in the fields `RecordSObjectClassification__c` and `RecordSObjectType__c`
4. **Fixed** - Data masking rules were not being applied to record collections (`List<SObject>`). Previously, data masking was only being applied when logging a single `SObject` record

**Comment by jongpie** [comment]
@vr8hub I've released [v4.6.9](https://github.com/jongpie/NebulaLogger/releases/tag/v4.7.9) last night, which includes these changes. Let me know if you run into any more issues, or need anything else!
